[issue-#6232] Added support for metrics signal in groupbyattrsprocessor

bertysentry opened the pull request:
**Description:**
Added support for *Metrics* in **groupbyattrsprocessor**

**Link to tracking Issue:**
Issue #6232 

**Testing:**
All unit tests have been updated to test metrics

**Documentation:**
README.md has been updated to mention support for metrics.

bertysentry committed:
[issue-#6232] **groupbyattrsprocessor** Added support for *Metrics* signal

* Added support for *Metrics* signal
* Fixed bug when overlapping attributes in parsed record and original *Resource*
* Added details and examples to **README.md**
* Some refactoring for clarification

MovieStoreGuy commented on processor/groupbyattrsprocessor/attribute_groups.go:
```diff
@@ -187,3 +225,21 @@ func (lgba *logsGroupedByAttrs) attributeGroup(baseResource pdata.Resource, reco
+// attributeGroup searches for a group with matching attributes and returns it. If nothing is found, it is being created
+func (mgba *metricsGroupedByAttrs) attributeGroup(baseResource pdata.Resource, recordAttrs pdata.AttributeMap) pdata.ResourceMetrics {
+	res, found := mgba.findGroup(baseResource, recordAttrs)
+	if !found {
+		res = mgba.AppendEmpty()
+		baseResource.CopyTo(res.Resource())
+
+		// This prioritizes metric attributes over resource attributes, if they overlap
+		attrs := res.Resource().Attributes()
+		recordAttrs.Range(func(k string, v pdata.AttributeValue) bool {
+			attrs.Upsert(k, v)
+			return true
+		})
+	}
+
```
Could we invert the condition here to reduce the amount of branching?

```suggestion
if found {
   return res
}
res = mgba.AppendEmpty()
baseResource.CopyTo(res.Resource())

// This prioritizes metric attributes over resource attributes, if they overlap
attrs := res.Resource().Attributes()
recordAttrs.Range(func(k string, v pdata.AttributeValue) bool {
    attrs.Upsert(k, v)
    return true
})
return res
```

bertysentry replied:
Okay, and I'll repeat the same pattern for Logs and Traces as well.

MovieStoreGuy commented on processor/groupbyattrsprocessor/processor.go:
```diff
@@ -133,3 +189,55 @@ func (gap *groupByAttrsProcessor) splitAttrMap(attrMap pdata.AttributeMap) (bool
+		// so we can delete those on the datapoint
+		deleteAttributes(groupedAttrMap, attributes)
+	} else {
+		stats.Record(ctx, mNumNonGroupedMetrics.M(1))
+	}
+
+	// Get the ResourceMetrics matching with these attributes
+	groupedResource := extractedGroups.attributeGroup(rm.Resource(), groupedAttrMap)
+
+	// Get the corresponding instrumentation library
+	groupedInstrumentationLibrary := matchingInstrumentationLibraryMetrics(groupedResource, ilm.InstrumentationLibrary())
+
+	// Return the metric in this resource
+	return getMetricInInstrumentationLibrary(groupedInstrumentationLibrary, metric)
+}
```
Be sure to add a new line here

bertysentry replied:
Okay! (Is this a convention?)

MovieStoreGuy reviewed (approved):
Just two things but nothing that would impact runtime.

bertysentry commented:
Rebased on **main** to make sure things will build smoothly. @MovieStoreGuy Could you please trigger the GitHub actions?

bertysentry commented:
@owais @anuraaga If you guys have some time to review this new feature, that would be awesome. It's already been tested to transform flat metrics from Prometheus to OTLP to DataDog (which leverages the concept of *Resources* to create corresponding hosts), and it works great in real use cases.

Aneurysm9 commented on processor/groupbyattrsprocessor/doc.go:
```diff
@@ -0,0 +1,17 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package filterprocessor implements a processor for filtering
+// (dropping) metrics and/or spans by various properties.
```
This package description doesn't look correct.

bertysentry replied:
My bad, thank you for pinpointing this!

bertysentry commented:
@anuraaga As code owner, please review, thank you in advance!

anuraaga commented on processor/groupbyattrsprocessor/processor_test.go:
```diff
@@ -318,3 +391,18 @@ func someLogs(attrs pdata.AttributeMap, count int) pdata.Logs {
+		metric := ilm.Metrics().AppendEmpty()
+		metric.SetName(fmt.Sprint("foo-", i))
+		metric.SetDataType(pdata.MetricDataTypeGauge)
+		dataPoint := metric.Gauge().DataPoints().AppendEmpty()
```
Not sure but looks like we're missing tests for non-gauge types? Or maybe I just didn't find them.

Also we definitely need a test with two metrics of different type but with same attributes, which won't be grouped together.

bertysentry replied:
Yeah you're right: I haven't implemented tests for the non-Gauge metric types, because I wanted to make sure that everybody was on the same page on the logic before going too far in unit tests (given that the code for sums and histograms is perfectly identical).

bertysentry replied:
I will update with more unit tests.

bertysentry replied:
Tests updated! (a lot more to cover different cases, which helped me squash an undetected bug)

anuraaga reviewed (commented):
I am somewhat worried about corner cases for metrics - for example is it ok to combine points from different collection intervals into the same Metric? I guess it is but I'm not sure if it could have some implications.

The code otherwise seems fine though, but looking at the linked issue, the motivation seems to be a problem with the Prometheus receiver populating resource information like host.name in metric instead of resource. Shouldn't the Prometheus receiver be fixed to fill in resource properly instead so users don't need a processor to fix it?

bertysentry commented:
@anuraaga I do not understand what scenario you imply when you say:
> for example is it ok to combine points from different collection intervals into the same Metric?

_DataPoints_ are not "combined". _DataPoints_ are simply copied from the original _Resource_ to the newly created _Resource_ with matching _Attributes_. So the origin and time of the _DataPoint_ really shouldn't matter, but I may miss something (as I'm very new to OpenTelemetry, so please forgive anything stupid I may say 😅)

Regarding the Prometheus Receiver, it is true that it could have configurable options to create _Resources_ as per some specified _Attributes_ keys found on the imported metrics, but that may be out of the scope of a receiver, especially since we have the **groupbyattrs** processor. 

And it doesn't seem logical to me that **groupbyattrs** would work on Traces and Logs, but not on Metrics. I still think this is a nice addition, given the flat nature of Prometheus metrics notably (but not only!).

Thank you!

anuraaga commented:
The scenario I think of, which I am not entirely sure is correct since I may be misreading, is that there are some data points in one Metric, and some in another Metric, that both have the same attributes. From what I could tell they will get merged into the same Metric. I don't know if this is the correct semantics for the Metric class.

 > Regarding the Prometheus Receiver, it is true that it could have configurable options to create Resources as per some specified Attributes keys found on the imported metrics

I don't know if it even needs to be configurable, isn't it more of a Prometheus data model mapping issue? I believe the spec is currently working on more robust mapping between the two this could be an issue to raise there. It may turn out that the receiver is exactly where this should be mapped in an official way.

> And it doesn't seem logical to me that groupbyattrs would work on Traces and Logs, but not on Metrics

I think it's because while Prometheus is flat, OTLP, which is the collector data model, isn't so the logic is not as clear (as above point). It took a lot of reading of the code to reason through the behavior and confirm that different data types wouldn't be merged into one. It's much more complex than spans and logs.

That doesn't mean there shouldn't be one but I don't think it's a trivial transformation. We need a thorough review from a metrics expert I think, maybe @bogdandrutu

bertysentry commented:
> The scenario [...] is that there are some data points in one Metric, and some in another Metric, that both have the same attributes. From what I could tell they will get merged into the same Metric.

To be "merged" into the same _Metric_ instance, the _DataPoints_ must have the same specified _Attributes_ **and also** belong to a _Metric_ with the same _Name_ and same _DataType_. In practice, you should never have 2 separate _Metric_ instances with the exact same name, unit and datatype (even though it doesn't make any difference, as they will be exported in the very same way).

Say you have the below metric points as input:

* _Resource_: `{host="my-exporter"}`
    * _Metric_: `my_metric_total` (Sum)
        * _DataPoint_: `{host="host-A",id="eth0"}` 1234
        * _DataPoint_: `{host="host-A",id="eth0"}` 1235
        * _DataPoint_: `{host="host-B",id="eth0"}` 0
        * _DataPoint_: `{host="host-B",id="eth0"}` 1
    * _Metric_: `other_metric` (Gauge)
        * _DataPoint_: `{host="host-A",id="eth0"}` 13
        * _DataPoint_: `{host="host-A",id="eth0"}` 2
        * _DataPoint_: `{host="host-B",id="eth0"}` 7
        * _DataPoint_: `{host="host-B",id="eth0"}` 1

And you configure **groupbyattrs** to group by the `host` _Attribute_. You will get the below metric points in the output:

* _Resource_: `{host="host-A"}`
    * _Metric_: `my_metric_total` (Sum)
        * _DataPoint_: `{id="eth0"}` 1234
        * _DataPoint_: `{id="eth0"}` 1235
    * _Metric_: `other_metric` (Gauge)
        * _DataPoint_: `{id="eth0"}` 13
        * _DataPoint_: `{id="eth0"}` 2
* _Resource_: `{host="host-B"}`
    * _Metric_: `my_metric_total` (Sum)
        * _DataPoint_: `{id="eth0"}` 0
        * _DataPoint_: `{id="eth0"}` 1
    * _Metric_: `other_metric` (Gauge)
        * _DataPoint_: `{id="eth0"}` 7
        * _DataPoint_: `{id="eth0"}` 1

BTW, the name of this *processor* is a bit misleading, because it's actually **splitting** _Metrics_ into several _Resources_, rather than **grouping** them.

> > Regarding the Prometheus Receiver, it is true that it could have configurable options to create Resources as per some specified Attributes keys found on the imported metrics
> 
> I don't know if it even needs to be configurable, isn't it more of a Prometheus data model mapping issue? I believe the spec is currently working on more robust mapping between the two this could be an issue to raise there. It may turn out that the receiver is exactly where this should be mapped in an official way.

There is no such concept in Prometheus' data model, like _Resources_ or _Hosts_, unfortunately. Prometheus is very flat, as you said, so the OpenTelemetry receiver for Prometheus will not be able to map certain Prometheus labels into OpenTelemetry _Resources_, I'm afraid.

> That doesn't mean there shouldn't be one but I don't think it's a trivial transformation. We need a thorough review from a metrics expert I think, maybe @bogdandrutu

I totally agree: I started working on this thinking "gonna be a piece of cake, just copy-paste the Log part and `replaceAll("Log", "Metric")`, but it was more complicated with the different data types and the _DataPoints_, which hold the _Attributes_ (and not the _Metrics_).

@bogdandrutu Your input is welcome, thank you!

anuraaga commented:
> To be "merged" into the same Metric instance, the DataPoints must have the same specified Attributes and also belong to a Metric with the same Name and same DataType.

Yeah - I just don't know if OTLP indicates that there can't be more than one Metric with the same name in a batch, and if so whether it's semantically equivalent to merge them. I suspect so, just don't think it's clear.

> There is no such concept in Prometheus' data model, like Resources or Hosts, unfortunately.

Yup, but I guess that there could still be a defacto mapping with pretty good coverage. For example, if `host.name`, `hostname`, `host-name`, `host` were all mapped to OTel host.name resource attribute, I suspect 99.9% of users would be covered, without having to add this processor. That would be a better UX in general. This is a hypothesis of course but if it seems conventional enough within Prometheus than I would definitely recommend it over having users to explicitly define it. It may not be conventional enough as my 99.9% number is made up. But it's my gut feeling.

And again that doesn't mean there shouldn't be a group by attr processor, but it needs to be clear of corner cases within OTLP's not-flat model, and I think its presence could be considered independent of a better mapping within the receiver.

bertysentry commented:
@anuraaga @MovieStoreGuy Please help!

The build is failing with this error:

```
Generated code is out of date, please run "make generate" and commit the changes in this PR.
Error: Process completed with exit code 1.
```

However, when I run `make generate` as advised in the error message, there is no change in the source, hence nothing that I can commit.

Any idea?

(Apart from that: I rebased the PR, squashed latest changes, fixed a bug, added unit tests as required and updated **README.md**, please review! 😉)

bertysentry committed:
[issue-#6232] Run `make goporto`

bertysentry commented:
@bogdandrutu @anuraaga @Aneurysm9 All tests are now successful. **groupbyattrs* processor now fully supports *Metrics* and documentation has been updated accordingly to clarify how the processor works.
Please review and approve if you guys are okay with the new feature.
Thank you in advance for your time!

codeboten reviewed (commented):
Pinging @pmm-sumo as well as the codeowner for the component

pmm-sumo reviewed (approved):
This is a very useful extension of the processor

bertysentry commented:
As everybody approved (@pmm-sumo @Aneurysm9 @MovieStoreGuy) and the build is all good, could someone please push the `Merge` button? @anuraaga ? Thank you!

MovieStoreGuy commented:
@bogdandrutu, @jpkrohling, or @tigrannajaryan ?

jpkrohling commented:
I added this to my queue. I trust the folks who reviewed this already, I'll just take a quick look before hitting the button.

jpkrohling commented on processor/groupbyattrsprocessor/README.md:
```diff
@@ -24,19 +95,28 @@ processors:
-* `num_grouped_logs` represents the number of logs that had attributes grouped
-* `num_non_grouped_logs` represents the number of logs that did not have attributes grouped
-* `log_groups` represents the distributon of groups extracted for logs
+| Metric | Description |
```
This table is so nice it deserves a better formatting:

```
| Metric                    | Description                                              |
|---------------------------|----------------------------------------------------------|
| `num_grouped_spans`       | the number of spans that had attributes grouped          |
| `num_non_grouped_spans`   | the number of spans that did not have attributes grouped |
| `span_groups`             | distributon of groups extracted for spans                |
| `num_grouped_logs`        | number of logs that had attributes grouped               |
| `num_non_grouped_logs`    | number of logs that did not have attributes grouped      |
| `log_groups`              | distributon of groups extracted for logs                 |
| `num_grouped_metrics`     | number of metrics that had attributes grouped            |
| `num_non_grouped_metrics` | number of metrics that did not have attributes grouped   |
| `metric_groups`           | distributon of groups extracted for metrics              |
```

jpkrohling commented on processor/groupbyattrsprocessor/README.md:
```diff
@@ -24,19 +95,28 @@ processors:
+|---|---|
+| `num_grouped_spans` | the number of spans that had attributes grouped |
+| `num_non_grouped_spans` | the number of spans that did not have attributes grouped |
+| `span_groups` | distributon of groups extracted for spans |
```
```suggestion
| `span_groups` | distribution of groups extracted for spans |
```

jpkrohling commented on processor/groupbyattrsprocessor/README.md:
```diff
@@ -24,19 +95,28 @@ processors:
+| `span_groups` | distributon of groups extracted for spans |
+| `num_grouped_logs` | number of logs that had attributes grouped |
+| `num_non_grouped_logs` | number of logs that did not have attributes grouped |
+| `log_groups` | distributon of groups extracted for logs |
```
```suggestion
| `log_groups` | distribution of groups extracted for logs |
```

jpkrohling commented on processor/groupbyattrsprocessor/README.md:
```diff
@@ -24,19 +95,28 @@ processors:
+| `log_groups` | distributon of groups extracted for logs |
+| `num_grouped_metrics` | number of metrics that had attributes grouped |
+| `num_non_grouped_metrics` | number of metrics that did not have attributes grouped |
+| `metric_groups` | distributon of groups extracted for metrics |
```
```suggestion
| `metric_groups` | distribution of groups extracted for metrics |
```

jpkrohling commented on processor/groupbyattrsprocessor/attribute_groups.go:
```diff
@@ -76,114 +97,150 @@ func newSpansGroupedByAttrs() *spansGroupedByAttrs {
+}
+
+// findResourceOrElseCreate searches for a Resource with matching attributes and returns it. If nothing is found, it is being created
+func (sgba *spansGroupedByAttrs) findResourceOrElseCreate(originResource pdata.Resource, requiredAttributes pdata.AttributeMap) pdata.ResourceSpans {
```
Perhaps `findOrCreateResource` would be a better name?

```suggestion
func (sgba *spansGroupedByAttrs) findOrCreateResource(originResource pdata.Resource, requiredAttributes pdata.AttributeMap) pdata.ResourceSpans {
```

jpkrohling commented on processor/groupbyattrsprocessor/factory.go:
```diff
@@ -104,7 +105,7 @@ func createTracesProcessor(
 }
 
-// createLogsProcessor creates a metrics processor based on this config.
+// createLogsProcessor creates a logs processor based on this config.
```
\o/

jpkrohling commented on processor/groupbyattrsprocessor/metrics.go:
```diff
@@ -24,9 +24,14 @@ var (
+
+	mNumGroupedLogs    = stats.Int64("num_grouped_logs", "Number of logs that had attributes grouped", stats.UnitDimensionless)
+	mNumNonGroupedLogs = stats.Int64("num_non_grouped_logs", "Number of logs that did not have attributes grouped", stats.UnitDimensionless)
+	mDistLogGroups     = stats.Int64("log_groups", "Distributon of groups extracted for logs", stats.UnitDimensionless)
```
```suggestion
	mDistLogGroups     = stats.Int64("log_groups", "Distribution of groups extracted for logs", stats.UnitDimensionless)
```

jpkrohling commented on processor/groupbyattrsprocessor/metrics.go:
```diff
@@ -24,9 +24,14 @@ var (
+
+	mNumGroupedMetrics    = stats.Int64("num_grouped_metrics", "Number of metrics that had attributes grouped", stats.UnitDimensionless)
+	mNumNonGroupedMetrics = stats.Int64("num_non_grouped_metrics", "Number of metrics that did not have attributes grouped", stats.UnitDimensionless)
+	mDistMetricGroups     = stats.Int64("metric_groups", "Distributon of groups extracted for metrics", stats.UnitDimensionless)
```
```suggestion
	mDistMetricGroups     = stats.Int64("metric_groups", "Distribution of groups extracted for metrics", stats.UnitDimensionless)
```

jpkrohling commented on processor/groupbyattrsprocessor/metrics.go:
```diff
@@ -24,9 +24,14 @@ var (
 	mNumGroupedSpans    = stats.Int64("num_grouped_spans", "Number of spans that had attributes grouped", stats.UnitDimensionless)
 	mNumNonGroupedSpans = stats.Int64("num_non_grouped_spans", "Number of spans that did not have attributes grouped", stats.UnitDimensionless)
 	mDistSpanGroups     = stats.Int64("span_groups", "Distributon of groups extracted for spans", stats.UnitDimensionless)
```
Would you mind fixing this one as well?

```suggestion
	mDistSpanGroups     = stats.Int64("span_groups", "Distribution of groups extracted for spans", stats.UnitDimensionless)
```

jpkrohling commented on processor/groupbyattrsprocessor/processor.go:
```diff
@@ -103,33 +103,152 @@ func (gap *groupByAttrsProcessor) processLogs(ctx context.Context, ld pdata.Logs
 }
 
+func (gap *groupByAttrsProcessor) processMetrics(ctx context.Context, md pdata.Metrics) (pdata.Metrics, error) {
+	rms := md.ResourceMetrics()
```
This requires more knowledge about metrics than what I currently have, I trust the other reviewers took care of this already :-)

jpkrohling reviewed (commented):
Looks good, just a couple of typos and minor comments.

bertysentry commented:
> Looks good, just a couple of typos and minor comments.

All reported typos and suggestions have been implemented. Thank you for the review!

bertysentry committed:
[issue-#6232] **groupbyattrs** Fixed a bunch of minor issues after code review

bertysentry commented:
@jpkrohling All checks have passed now, so we're good to go. Thank you!

bertysentry commented:
Yay! 😊
